### PR TITLE
Refactor ConvertTimeMajorSpectrogram kernel

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

## Category:

**Refactoring**

## Description:

Divides ConvertTimeMajorSpectrogram into 3 kernels
- one for in-place and different type width
- one for the general case
- one when there is no padding (shape can be flattened)
It  boosts performance of its typical usage in Spectrogram op by about 12%

### Affected modules and functionalities:

Spectrogram op

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-3082

### CODE REVIEW

File: dali/kernels/signal/fft/fft_postprocess.cuh [hunk 1]
```diff
@@ -110,38 +110,58 @@ struct power_dB {
...
+template <typename Out, typename In, typename Convert>
+__global__ void ConvertTimeMajorSpectrogram_InPlaceDiffTypeSize(
+      Out *out, int out_stride,
+      const In *in, int in_stride, int nfft, int64_t nwindows,
+      Convert convert = {}) {
   // A warp processes a whole row (transform) to ensure sequential processing
   // and thus enable in-place operation.
 
-  int64_t wnd = static_cast<int64_t>(blockIdx.y) * blockDim.y + threadIdx.y;
+  int64_t wnd = static_cast<int64_t>(blockIdx.x) * blockDim.y + threadIdx.y;
```
Comment (jantonguirao) [inline dali/kernels/signal/fft/fft_postprocess.cuh:137]: using `x` for the grid so that I can use a larger value

**Review by mzient — APPROVED** [review]

File: dali/kernels/signal/fft/fft_postprocess.cuh [hunk 2]
```diff
@@ -110,38 +110,58 @@ struct power_dB {
...
+    return;
+  out += wnd * out_stride;
+  in += wnd * in_stride;
+
+  for (int i = threadIdx.x; i < nfft; i += 32)
+    out[i] = convert(in[i]);
+}
+
+template <typename Out, typename In, typename Convert>
+__global__ void ConvertTimeMajorSpectrogram_InPlaceDiffTypeSize(
```
Comment (klecki) [inline dali/kernels/signal/fft/fft_postprocess.cuh:130]: Until I found that we are only using it with float and float2 it looked a bit like UB.

**Review by klecki — COMMENTED** [review]
The code looks like it does what it is supposed to do (extracting the if from device to host), posted small nitpicks.

Do we have tests that cover all three code-paths? The PR description doesn't indicate any tests.

**Comment by jantonguirao** [comment]
!build

**Review by klecki — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

**Pull request merged**